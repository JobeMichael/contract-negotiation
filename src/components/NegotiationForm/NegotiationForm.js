import React, { useContext, useEffect } from "react";
import isDataSubmitted from "../../helper/isDataSubmitted";
import validate from "../../helper/validation";
import useForm from "../../hooks/useForm";
import { appContext } from "../../store/appContext";
import Button from "../UI/Button/Button";
import Input from "../UI/Input/Input";
import * as Styled from "./NegotiationForm.styles";

const NegotiationForm = ({ setOpenResultCard }) => {
  const {
    appState: { currentUser, employerInput, employeeInput },
  } = useContext(appContext);
  const { values, errors, handleChange, handleSubmit } = useForm(validate);

  useEffect(() => {
    if (employeeInput > 0 && employerInput > 0) {
      setOpenResultCard(true);
    }
  }, [employeeInput, employerInput, setOpenResultCard]);

  const hideField = isDataSubmitted(employerInput, employeeInput, currentUser);

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <Styled.FieldWrapper>
          {hideField ? (
            <Styled.InfoPanel>Data submitted</Styled.InfoPanel>
          ) : (
            <>
              <Input
                name={currentUser}
                type="number"
                label="Amount"
                value={values[currentUser] || ""}
                onChange={handleChange}
                errorMsg={errors[currentUser]}
                data-testid="input-field"
              />
              <Button>Submit</Button>
            </>
          )}
        </Styled.FieldWrapper>
      </form>
    </div>
  );
};

export default NegotiationForm;
