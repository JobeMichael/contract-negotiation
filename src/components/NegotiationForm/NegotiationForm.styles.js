import styled from "styled-components";

export const FieldWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 40px 0;
  box-sizing: border-box;
  border-bottom: none;
`;

export const InfoPanel = styled.p`
  color: green;
  font-size: 20px;
  font-weight: 600;
  margin-top: 40px;
`;
