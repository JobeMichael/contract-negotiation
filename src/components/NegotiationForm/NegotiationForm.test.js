import { act, cleanup, fireEvent, render } from "@testing-library/react";
import React from "react";
import { appContext } from "../../store/appContext";
import NegotiationForm from "./NegotiationForm";

describe("<NegotiationForm/>", () => {
  const setOpenResultCard = jest.fn();
  const setAppState = jest.fn();

  afterEach(() => {
    cleanup();
    setOpenResultCard.mockRestore();
    setAppState.mockRestore();
  });

  let renderResult;

  const setup = (context) => (
    <appContext.Provider value={context}>
      <NegotiationForm setOpenResultCard={setOpenResultCard} />
    </appContext.Provider>
  );

  it("renders NegotiationForm properly", () => {
    const mockContext = {
      appState: {
        currentUser: "employer",
        employerInput: 0,
        employeeInput: 0,
      },
    };
    act(() => {
      renderResult = render(setup(mockContext));
    });

    expect(renderResult.baseElement).toMatchSnapshot();
  });

  it("call setOpenResultCard after successfully submit data", () => {
    const mockContext = {
      appState: {
        currentUser: "employer",
        employerInput: 1,
        employeeInput: 1,
      },
    };
    act(() => {
      renderResult = render(setup(mockContext));
    });

    expect(setOpenResultCard).toBeCalledTimes(1);
  });

  it("should change input fields value", () => {
    const mockContext = {
      appState: {
        currentUser: "employer",
        employerInput: 0,
        employeeInput: 0,
      },
    };

    act(() => {
      renderResult = render(setup(mockContext));
    });

    const employer = renderResult.baseElement.querySelector(
      "input[name='employer']"
    );

    fireEvent.change(employer, {
      target: { value: "1234" },
    });
    expect(employer.value).toBe("1234");
  });

  it("should submit the form and update context", () => {
    const mockContext = {
      appState: {
        currentUser: "employer",
        employerInput: 0,
        employeeInput: 0,
      },
      setAppState: setAppState,
    };

    act(() => {
      renderResult = render(setup(mockContext));
    });

    const employer = renderResult.baseElement.querySelector(
      "input[name='employer']"
    );
    fireEvent.change(employer, {
      target: { value: "1234" },
    });
    const submit = renderResult.baseElement.querySelector("button");
    act(() => {
      fireEvent.click(submit);
    });

    expect(setAppState).toBeCalledTimes(1);
  });
});
