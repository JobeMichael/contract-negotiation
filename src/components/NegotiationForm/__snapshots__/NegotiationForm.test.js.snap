// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NegotiationForm/> renders NegotiationForm properly 1`] = `
<body>
  <div>
    <div>
      <form>
        <div
          class="sc-AxiKw bJLBAq"
        >
          <div
            class="sc-AxirZ iYVKFQ"
          >
            <label>
              Amount
            </label>
            <div>
              <input
                autocomplete="off"
                class="input-element"
                data-testid="input-field"
                name="employer"
                type="number"
                value=""
              />
            </div>
          </div>
          <button
            class="sc-AxjAm chZpQe button"
            data-testid="button-submit"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</body>
`;
